src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h

```diff
@@ -163,12 +172,43 @@ Node RewriteRule<CompEliminate>::apply(TNode node)
 }
 
 template <>
-bool RewriteRule<SubEliminate>::applies(TNode node) {
+inline bool RewriteRule<AshrEliminate>::applies(TNode node)
+{
+  return (node.getKind() == kind::BITVECTOR_ASHR);
+}
+
+template <>
+inline Node RewriteRule<AshrEliminate>::apply(TNode node)
+{
+  Debug("bv-rewrite") << "RewriteRule<AshrEliminate>(" << node << ")"
+                      << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
+  TNode s = node[0];
+  TNode t = node[1];
+  /*  From SMT-LIB:
+   *  (bvashr s t) abbreviates
+   *    (ite (= ((_ extract |m-1| |m-1|) s) #b0)
+   *         (bvlshr s t)
+   *         (bvnot (bvlshr (bvnot s) t)))  */
```

Comment:
A problem with using this directly is that it introduces two shifts and so doubles the size of the bit-blasting.  That is a significant performance overhead.  If we must do this rewrite, can we at least:
```
(let mask
      ((_ repeat m) ((_ extract |m-1| |m-1|) s))
      (bvxor mask (bvlshr (bvxor mask s) t)))
```

Reply:
Since the only purpose of this rewrite rules is to make the bv-to-int preprocessing pass easier, this optimization actually makes that translation harder. First, `bvxor` is harder to handle in integers than the alternatives. Second, it is also convenient to assume that `repeat` does not occur when translating to integers :)

Reply:
I agree with @martin-cs, if it's better for your preprocessing pass to eliminate `bvashr` in this way, it's probably better to move the elimination into your preprocessing pass. In general, if we have to eliminate `bvashr` I would prefer @martin-cs's encoding.

Reply:
Sounds good. I will eliminate `bvashr` internally in my pass. For now, I have changed this PR to a small refactoring PR, that adds the inclusion of the options header file and adds `inline` to the functions.